### PR TITLE
Migrating Push Gossip to SDK handlers

### PR DESCRIPTION
**Pull request by joshua-kim**
## Why this should be merged

Continuation on our work to standardize our p2p logic, this moves all of our push gossip workflow into SDK handlers.

## How this works

Adds handlers for push gossip into the SDK's router. Following the next network upgrade, we'll deprecate the old gossip gossip sent without the SDK.

## How this was tested

Added a unit test

### CODE REVIEW

File: plugin/evm/tx_gossip_test.go [hunk 1]
```diff
@@ -5,250 +5,308 @@ package evm
...
-	"go.uber.org/mock/gomock"
-
-	"google.golang.org/protobuf/proto"
-
 	"github.com/ava-labs/coreth/core"
 	"github.com/ava-labs/coreth/core/types"
 	"github.com/ava-labs/coreth/params"
 )
 
 func TestEthTxGossip(t *testing.T) {
```
Comment (joshua-kim) [inline plugin/evm/tx_gossip_test.go:30]: The old tests were unbelievably confusing to follow. The new tests just set up two vms, issue a tx to one of them, and wait until we see it.

File: plugin/evm/vm_test.go [hunk 2]
```diff
@@ -124,6 +124,69 @@ func init() {
...
+		}
+	}
+
+	genesis := newGenesis()
+	genesis.Alloc = alloc
+
+	return genesis
+}
+
+func newGenesis() *core.Genesis {
```
Comment (joshua-kim) [inline plugin/evm/vm_test.go:144]: I feel like the way we're testing the VM by passing in a string-ed json is confusing.... added some code to just pass around a genesis struct instead + initialize a funded genesis instead of imports w/ shared memory.

File: plugin/evm/gossiper_atomic_gossiping_test.go [hunk 3]
```diff
@@ -11,75 +11,12 @@ import (
...
 	"github.com/ava-labs/avalanchego/ids"
-	"github.com/ava-labs/avalanchego/snow"
 	"github.com/ava-labs/avalanchego/utils/set"
 	"github.com/stretchr/testify/assert"
 
 	"github.com/ava-labs/coreth/plugin/evm/message"
 )
 
-// locally issued txs should be gossiped
-func TestMempoolAtmTxsIssueTxAndGossiping(t *testing.T) {
```
Comment (joshua-kim) [inline plugin/evm/gossiper_atomic_gossiping_test.go:22]: This seems to test the same behavior as the new atomic tx test so it's deleted.

File: plugin/evm/vm.go [hunk 4]
```diff
@@ -1,4 +1,3 @@
-// (c) 2019-2020, Ava Labs, Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (joshua-kim) [reply plugin/evm/vm.go]: oops :<

File: plugin/evm/vm_test.go [hunk 5]
```diff
@@ -124,6 +118,69 @@ func init() {
...
+			ApricotPhase4BlockTimestamp:     new(uint64),
+			ApricotPhase5BlockTimestamp:     new(uint64),
+			ApricotPhasePre6BlockTimestamp:  new(uint64),
+			ApricotPhase6BlockTimestamp:     new(uint64),
+			ApricotPhasePost6BlockTimestamp: new(uint64),
+			BanffBlockTimestamp:             new(uint64),
+			CortinaBlockTimestamp:           new(uint64),
+			DUpgradeBlockTimestamp:          new(uint64),
+			CancunTime:                      new(uint64),
+		},
```
[the remark this answers is not in the document]
Comment (joshua-kim) [reply plugin/evm/vm_test.go]: Hmm I didn't know this package existed, I think that makes sense to use. I think it would still be a good idea to have a function that creates these configs per test as opposed to having a global variable in case tests want to override a default without affecting the state of that config for other tests.

File: plugin/evm/vm.go [hunk 6]
```diff
@@ -1023,92 +1019,127 @@ func (vm *VM) initBlockBuilding() error {
...
+	gossipStats := NewGossipStats()
+	vm.gossipHandler = NewGossipHandler(vm, gossipStats)
+
+	ethTxPushGossipHandler := ethTxPushGossipHandler{
+		NoOpHandler:   p2p.NoOpHandler{},
+		GossipHandler: vm.gossipHandler,
+		codec:         vm.networkCodec,
+	}
+
+	ethTxPushGossipClient, err := vm.router.RegisterAppProtocol(ethTxPushGossipProtocol, ethTxPushGossipHandler, nil)
```
Comment (aaronbuchwald) [inline plugin/evm/vm.go:1122]: Can we pass in a non-nil sampler here? Looking at the p2p SDK it looks like this will panic if we ever send an `AppRequestAny` on this type.

We should also improve the documentation around this if the client is going to pass in a sampler, since the documentation for `AppRequestAny` and `RegisterAppProtocol` do not mention that this combination could cause a panic.

File: plugin/evm/vm_test.go [hunk 7]
```diff
@@ -4092,3 +4104,282 @@ func TestSkipChainConfigCheckCompatible(t *testing.T) {
...
+			require.False(vm2.txPool.Has(signedTx.Hash()))
+
+			require.NoError(vm1.txPool.AddRemote(signedTx))
+			require.True(vm1.txPool.Has(signedTx.Hash()))
+
+			update := make(chan core.NewTxsEvent)
+			vm2.txPool.SubscribeNewTxsEvent(update)
+			<-update
+
+			require.True(vm2.txPool.Has(signedTx.Hash()))
```
Comment (aaronbuchwald) [inline plugin/evm/vm_test.go:4219]: ❤️ this is great

Comment (aaronbuchwald) [reply plugin/evm/vm_test.go:4219]: Not necessary for this PR, but it would be great to factor out this structure from these two tests so that we can easily re-use this to test vm2vm networking behaviors.

**Review by aaronbuchwald — DISMISSED** [review]

File: plugin/evm/vm.go [hunk 8]
```diff
@@ -1023,92 +1019,127 @@ func (vm *VM) initBlockBuilding() error {
...
+	if err != nil {
+		return err
+	}
+
+	atomicTxPushGossipHandler := atomicTxPushGossipHandler{
+		NoOpHandler:   p2p.NoOpHandler{},
+		GossipHandler: vm.gossipHandler,
+		codec:         vm.networkCodec,
+	}
```
Comment (ceyonur) [inline plugin/evm/vm.go:1131]: Seems there are bunch of duplicated code. I wonder if we can combine them into a single function that works with interfaces and take handlers as parameter to return client. I.e initGossiper(mempool gossip.gossippable, protocol, gossip.config...)

File: plugin/evm/vm.go [hunk 9]
```diff
@@ -1023,92 +1019,127 @@ func (vm *VM) initBlockBuilding() error {
...
+
+	atomicTxPushGossipClient, err := vm.router.RegisterAppProtocol(atomicTxPushGossipProtocol, atomicTxPushGossipHandler, nil)
+	if err != nil {
+		return err
+	}
+
+	vm.gossiper = vm.createGossiper(gossipStats, ethTxPushGossipClient, atomicTxPushGossipClient)
+	vm.builder = vm.NewBlockBuilder(vm.toEngine)
+	vm.builder.awaitSubmittedTxs()
+	vm.Network.SetGossipHandler(vm.gossipHandler)
```
Comment (ceyonur) [inline plugin/evm/vm.go:1141]: @aaronbuchwald do you think we should first set the handler then call `awaitSubmittedTxs`?
This is unrelated with this PR.

File: plugin/evm/vm.go (hunk 6 above)
Comment (joshua-kim) [reply plugin/evm/vm.go:1122]: Fixed w/ the new refactor

File: plugin/evm/vm.go [hunk 10]
```diff
@@ -1023,92 +1019,127 @@ func (vm *VM) initBlockBuilding() error {
...
-		gossip.Every(ctx, vm.ctx.Log, atomicTxGossiper, gossipFrequency)
+		gossip.Every(ctx, vm.ctx.Log, atomicTxPullGossiper, gossipFrequency)
 		vm.shutdownWg.Done()
 	}()
 
+	gossipStats := NewGossipStats()
+	vm.gossipHandler = NewGossipHandler(vm, gossipStats)
+
+	ethTxPushGossipHandler := ethTxPushGossipHandler{
+		NoOpHandler:   p2p.NoOpHandler{},
```
[the remark this answers is not in the document]
Comment (joshua-kim) [reply plugin/evm/vm.go]: Added

File: plugin/evm/gossiper.go [hunk 11]
```diff
@@ -43,6 +45,8 @@ const (
 	minGossipBatchInterval = 50 * time.Millisecond
 )
 
+var _ p2p.Handler = (*ethTxPushGossipHandler)(nil)
```
[the remark this answers is not in the document]
Comment (joshua-kim) [reply plugin/evm/gossiper.go]: Removed this type

File: plugin/evm/gossiper.go [hunk 12]
```diff
@@ -300,8 +308,19 @@ func (n *pushGossiper) gossipAtomicTx(tx *Tx) error {
 		"gossiping atomic tx",
 		"txID", txID,
 	)
+
+	if err := n.atomicTxClient.AppGossip(context.TODO(), msgBytes); err != nil {
+		return err
+	}
+
 	n.stats.IncAtomicGossipSent()
```
[the remark this answers is not in the document]
Comment (joshua-kim) [reply plugin/evm/gossiper.go:331]: Added client-side metrics for the sdk gossipers

File: peer/network.go [hunk 13]
```diff
@@ -444,11 +444,11 @@ func (n *network) Gossip(gossip []byte) error {
 // AppGossip is called by avalanchego -> VM when there is an incoming AppGossip from a peer
 // error returned by this function is expected to be treated as fatal by the engine
 // returns error if request could not be parsed as message.Request or when the requestHandler returns an error
```
[the remark this answers is not in the document]
Comment (joshua-kim) [reply peer/network.go:451]: Updated comment

File: plugin/evm/vm.go (hunk 8 above)
Comment (joshua-kim) [reply plugin/evm/vm.go]: I think there should be less duplicated code now that there's only one gossip handler per gossip protocol... previously there was 2x2=4 which was definitely too much...

File: peer/network.go [hunk 14]
```diff
@@ -444,11 +444,11 @@ func (n *network) Gossip(gossip []byte) error {
...
-func (n *network) AppGossip(_ context.Context, nodeID ids.NodeID, gossipBytes []byte) error {
+func (n *network) AppGossip(ctx context.Context, nodeID ids.NodeID, gossipBytes []byte) error {
 	var gossipMsg message.GossipMessage
 	if _, err := n.codec.Unmarshal(gossipBytes, &gossipMsg); err != nil {
-		log.Debug("could not parse app gossip", "nodeID", nodeID, "gossipLen", len(gossipBytes), "err", err)
-		return nil
+		log.Debug("forwarding AppGossip to SDK router", "nodeID", nodeID, "gossipLen", len(gossipBytes), "err", err)
```
[the remark this answers is not in the document]
Comment (joshua-kim) [reply peer/network.go:455]: This is consistent w/ the current logs... going to leave this as-is for now

File: plugin/evm/vm.go [hunk 15]
```diff
@@ -1084,7 +1090,13 @@ func (vm *VM) initBlockBuilding() error {
 
 	if vm.ethTxGossipHandler == nil {
 		var ethTxGossipHandler p2p.Handler
-		ethTxGossipHandler, err = gossip.NewHandler[*GossipEthTx](ethTxPool, ethTxGossipHandlerConfig, vm.sdkMetrics)
+		ethTxGossipHandler = gossip.NewHandler[GossipEthTx, *GossipEthTx](
+			vm.ctx.Log,
+			vm.ethTxPushGossiper,
```
Comment (StephenButtolph) [inline plugin/evm/vm.go:1095]: We need to allow inbound `AppGossip` traffic from non-validators for both eth and atomic txs. Otherwise non-validators would have no way of issuing a tx.

File: plugin/evm/gossiper.go [hunk 16]
```diff
@@ -259,7 +270,24 @@ func (n *pushGossiper) awaitEthTxGossip() {
...
+					continue
+				}
+
+				gossipTxs := make([]*GossipEthTx, 0, len(txs))
+				for _, tx := range txs {
+					gossipTxs = append(gossipTxs, &GossipEthTx{Tx: tx})
+				}
+
+				n.ethTxGossiper.Add(gossipTxs...)
+				if err := n.ethTxGossiper.Gossip(context.TODO()); err != nil {
```
Comment (StephenButtolph) [inline plugin/evm/gossiper.go:282]: No sure if this is a valid concern - but it's possible that we added more transactions than we sent here. Could this ever result in a memory buildup?

File: plugin/evm/mempool.go [hunk 17]
```diff
@@ -152,6 +157,10 @@ func (m *Mempool) AddTx(tx *Tx) error {
 	defer m.lock.Unlock()
```
Comment (StephenButtolph) [inline plugin/evm/mempool.go:157]: addeed -> added? (couple lines above this)

Comment (joshua-kim) [reply plugin/evm/mempool.go:157]: Re-spelled

File: plugin/evm/gossiper.go (hunk 16 above)
Comment (joshua-kim) [reply plugin/evm/gossiper.go:282]: Hmm I guess this can happen if txs are coming in at a rate quicker than they're being gossiped.... but I feel like this is not too concerning since currently we're pretty much running a gossip cycle every time we see a tx I don't really expect this the txs waiting to be gossiped to build up to a noticeable size

File: plugin/evm/vm.go (hunk 15 above)
Comment (joshua-kim) [reply plugin/evm/vm.go:1094]: Should be fixed in the latest commit: https://github.com/ava-labs/coreth/pull/363/commits/d28cea07c33d5a299d13bf4e07649e500d767e56

File: plugin/evm/vm.go (hunk 8 above)
Comment (joshua-kim) [reply plugin/evm/vm.go]: Added function to do this

File: plugin/evm/gossip.go [hunk 18]
```diff
@@ -19,11 +22,65 @@ import (
...
+	throttlingPeriod time.Duration,
+	throttlingLimit int,
+	validators *p2p.Validators,
+) txGossipHandler {
+	// push gossip messages can be handled from any peer
+	handler := gossip.NewHandler[T, U](
+		log,
+		// Don't forward gossip to avoid double-forwarding while legacy
+		// gossip is still running.
+		gossip.NoOpAccumulator[U]{},
```
Comment (joshua-kim) [inline plugin/evm/gossip.go:46]: So I think for now we can't actually use the push gossiper to forward transactions. The mempool gets notified of a new tx by looking at a channel which is updated on add, but since this handler adds the new txs to the mempool this will result in a double gossip if we don't skip the push gossip step after adding to the mempool.

File: plugin/evm/gossiper.go [hunk 19]
```diff
@@ -259,7 +270,24 @@ func (n *pushGossiper) awaitEthTxGossip() {
 						"len(txs)", attempted,
 						"err", err,
 					)
+					continue
```
Comment (StephenButtolph) [inline plugin/evm/gossiper.go:273]: I don't think we actually need to `continue` here. But this `warn` shouldn't be firing anyways.. So it seems fine.

File: plugin/evm/gossiper.go [hunk 20]
```diff
@@ -259,7 +270,24 @@ func (n *pushGossiper) awaitEthTxGossip() {
...
+				}
+
+				n.ethTxGossiper.Add(gossipTxs...)
+				if err := n.ethTxGossiper.Gossip(context.TODO()); err != nil {
+					log.Warn(
+						"failed to send eth transactions",
+						"len(txs)", len(txs),
+						"err", err,
+					)
+					continue
```
Comment (StephenButtolph) [inline plugin/evm/gossiper.go:288]: This `continue` seems useless right?
```suggestion
```

File: plugin/evm/gossiper.go (hunk 16 above)
Comment (StephenButtolph) [reply plugin/evm/gossiper.go:282]: We could add a call to `Gossip` inside of the `gossipTicker` call if we wanted to further protect against this.

File: plugin/evm/mempool.go [hunk 21]
```diff
@@ -145,13 +150,17 @@ func (m *Mempool) Add(tx *GossipAtomicTx) error {
...
+// AddTx attempts to add [tx] to the mempool and returns an error if
+// it could not be added to the mempool.
 func (m *Mempool) AddTx(tx *Tx) error {
 	m.lock.Lock()
 	defer m.lock.Unlock()
 
 	err := m.addTx(tx, false)
+	if errors.Is(err, errTxAlreadyKnown) {
+		return nil
+	}
```
Comment (StephenButtolph) [inline plugin/evm/mempool.go:162]: We call `addTx` in `AddLocalTx` and `ForceAddTx` - don't we need to check for this there as well?

File: plugin/evm/vm.go [hunk 22]
```diff
@@ -144,39 +144,22 @@ const (
 	txGossipBloomMaxItems          = 8 * 1024
 	txGossipBloomFalsePositiveRate = 0.01
 	txGossipMaxFalsePositiveRate   = 0.05
-	txGossipTargetResponseSize     = 20 * units.KiB
+	txGossipMaxSize                = 20 * units.KiB
```
Comment (StephenButtolph) [inline plugin/evm/vm.go:147]: I don't think `maxSize` is correct - we allow larger sizes.

File: plugin/evm/gossiper.go (hunk 20 above)
Comment (joshua-kim) [reply plugin/evm/gossiper.go:288]: It is a no-op, it's just there because I feel like we might add code after this line in the future and unintentionally run it if this errors.

Comment (joshua-kim) [reply plugin/evm/gossiper.go:288]: It's a no-op.... I left it in because i was worried that there might be a world where we add more code and an error would result in stuff running that we didn't intend to run. I am fine removing this though

File: plugin/evm/mempool.go (hunk 21 above)
Comment (joshua-kim) [reply plugin/evm/mempool.go:162]: Yeah we do...

File: plugin/evm/vm.go (hunk 22 above)
Comment (joshua-kim) [reply plugin/evm/vm.go:147]: renamed to txGossipTargetMessageSize

File: plugin/evm/gossiper.go (hunk 16 above)
Comment (joshua-kim) [reply plugin/evm/gossiper.go:281]: I'll add this in.

File: plugin/evm/gossip.go [hunk 23]
```diff
@@ -19,11 +22,65 @@ import (
...
+	maxMessageSize int,
+	throttlingPeriod time.Duration,
+	throttlingLimit int,
+	validators *p2p.Validators,
+) txGossipHandler {
+	// push gossip messages can be handled from any peer
+	handler := gossip.NewHandler[T, U](
+		log,
+		// Don't forward gossip to avoid double-forwarding while legacy
+		// gossip is still running.
```
[the remark this answers is not in the document]
Comment (joshua-kim) [reply plugin/evm/gossip.go:45]: Yeah i can reword this to indicate that we're already forwarding so we don't need to double-forward here.

**Review by StephenButtolph — APPROVED** [review]

File: plugin/evm/mempool.go [hunk 24]
```diff
@@ -21,7 +21,12 @@ const (
 	discardedTxsCacheSize = 50
 )
 
-var errNoGasUsed = errors.New("no gas used")
+var (
+	errTxAlreadyKnown = errors.New("tx already known")
```
Comment (darioush) [inline plugin/evm/mempool.go:25]: why do we add this error then bypass it at all the callsites?

**Review by darioush — APPROVED** [review]

File: plugin/evm/mempool.go (hunk 24 above)
Comment (StephenButtolph) [reply plugin/evm/mempool.go:25]: You're totally right - this should be returned from `Add`.

Comment (StephenButtolph) [reply plugin/evm/mempool.go:25]: 🦅 👁️

Comment (joshua-kim) [reply plugin/evm/mempool.go:25]: Yeah this is supposed to be returned from `Add` to indicate that a duplicate existed...

File: plugin/evm/mempool.go [hunk 25]
```diff
@@ -142,16 +147,20 @@ func (m *Mempool) Add(tx *GossipAtomicTx) error {
 	m.ctx.Lock.RLock()
 	defer m.ctx.Lock.RUnlock()
 
-	return m.AddTx(tx.Tx)
+	return m.addTx(tx.Tx, false)
```
Comment (StephenButtolph) [inline plugin/evm/mempool.go:150]: This needs to hold the mempool's lock

Comment (StephenButtolph) [inline plugin/evm/mempool.go:150]: I think we should be marking the tx as discarded if it returns an error (other than `errTxAlreadyKnown`) right?

File: plugin/evm/mempool.go [hunk 26]
```diff
@@ -139,19 +144,40 @@ func (m *Mempool) atomicTxGasPrice(tx *Tx) (uint64, error) {
 }
 
 func (m *Mempool) Add(tx *GossipAtomicTx) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
```
Comment (StephenButtolph) [inline plugin/evm/mempool.go:148]: This grabs the locks in the wrong order. The context lock must be grabbed first

Comment (joshua-kim) [reply plugin/evm/mempool.go:148]: I'm still trying to figure out why we acquire the context lock here

File: plugin/evm/mempool.go [hunk 27]
```diff
@@ -139,19 +144,40 @@ func (m *Mempool) atomicTxGasPrice(tx *Tx) (uint64, error) {
...
 
-	return m.AddTx(tx.Tx)
+	err := m.addTx(tx.Tx, false)
+	if errors.Is(err, errTxAlreadyKnown) {
+		return err
+	}
+
+	if err != nil {
+		txID := tx.Tx.ID()
+		m.discardedTxs.Put(tx.Tx.ID(), tx.Tx)
```
Comment (StephenButtolph) [inline plugin/evm/mempool.go:160]:
```suggestion
		txID := tx.Tx.ID()
		m.discardedTxs.Put(txID, tx.Tx)
```

File: plugin/evm/mempool.go (hunk 26 above)
Comment (joshua-kim) [reply plugin/evm/mempool.go]: for future readers this is to lock state while `vm.verifyAtTip` is being run.

**Review by StephenButtolph — APPROVED** [review]
take 2
